flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionDirectTransferReader.java

```diff
@@ -91,10 +91,14 @@ public BufferAndBacklog getNextBuffer() throws IOException {
 
         updateStatistics(current);
 
-        // We simply assume all the data are non-events for batch jobs to avoid pre-fetching the
-        // next header
-        Buffer.DataType nextDataType =
-                numDataAndEventBuffers > 0 ? Buffer.DataType.DATA_BUFFER : Buffer.DataType.NONE;
+        // We simply assume all the data except for the last one (EndOfPartitionEvent)
+        // are non-events for batch jobs to avoid pre-fetching the next header
+        Buffer.DataType nextDataType = Buffer.DataType.NONE;
+        if (numDataBuffers > 0) {
+            nextDataType = Buffer.DataType.DATA_BUFFER;
+        } else if (numDataAndEventBuffers > 0) {
+            nextDataType = Buffer.DataType.EVENT_BUFFER;
+        }
```

Comment:
An EVENT_BUFFER does not need any credit to be sent. The BoundedBlockingSubpartitionDirectTransferReader Implementation does not distinguish EVENT_BUFFER and DATA_BUFFER for simplicity. However, after this change, we need to distinguish EVENT_BUFFER and DATA_BUFFER because we do not want to allocate any credit at downstream for an EVENT_BUFFER.

Reply:
But what would be a problem with requesting for a credit for the `EndOfPartitionEvent`? In other words, what's wrong with doing it as it was done previously: always returning `DATA_BUFFER` or `NONE`?

Reply:
I think this is just an optimization. Without this change, the backlog announced to the downstream does not include the event and EVENT_BUFFER and DATA_BUFFER are not distinguished, this means the event buffer also need a credit to send. If there is no exclusive buffer, no enough buffer will be allocated for the event, because we only announce the data buffer backlog to the downstream. As a result, some tests will hang for there is no credit for the event buffer. To solve this dead lock, there are two simple ways:
1. Announce both data buffer and event buffer backlog to the downstream, this lead to allocate more buffers than needed, these buffers will be released when the EOF is received at the downstream task.
2. Distinguish EVENT_BUFFER and DATA_BUFFER just like what is doing now.

These choices are both acceptable to me. I chose the second one because EVENT_BUFFER and DATA_BUFFER are distinguished at the downstream task and we can allocate one less buffer.

Both of the choices need to do some change to BoundedBlockingSubpartitionDirectTransferReader. Which one do you prefer? I think both are acceptable for me.

Reply:
After rethink about it, the first choice can support more events type in the future and the second choice make the assumption that we only have one event at the end of the data. Maybe the first choice is better?

Reply:
Let's stay with the option 2 (as you have currently implemented)